### PR TITLE
Feature/apmi 2313 implement disk caching

### PR DESCRIPTION
**Pull request by shattimare**
Implemented code.

There are two points to clarify.
1. When we generate span from cache data, we are not able to use cached value of span id , traceid, traceFlags, traceState, parentSpanId to export it on RUM playground. Reason of it is that setter method of that property is private, so we are not able to set it. It will randomly generate spanID, traceID etc.
2. After network will available , cached spans will not be exported automatically on playground instead cached span will be exported to playground with span which will generate by user action.

### CODE REVIEW

File: SplunkRumWorkspace/SplunkRum/SplunkRum/SplunkRum.swift [hunk 1]
```diff
@@ -180,7 +180,7 @@ var splunkRumInitializeCalledTime = Date()
         let limiting = LimitingExporter(proxy: retry, spanFilter: options?.spanFilter ?? nil)
         OpenTelemetrySDK.instance.tracerProvider.addSpanProcessor(BatchSpanProcessor(spanExporter: limiting))
         if options?.debug ?? false {
-            OpenTelemetrySDK.instance.tracerProvider.addSpanProcessor(SimpleSpanProcessor(spanExporter: StdoutExporter(isDebug: true)))
+            OpenTelemetrySDK.instance.tracerProvider.addSpanProcessor(SimpleSpanProcessor(spanExporter: StdoutExporter(isDebug: false)))
```
[the remark this answers is not in the document]
Comment (shattimare) [reply SplunkRumWorkspace/SplunkRum/SplunkRum/SplunkRum.swift:183]: isDebug should be true

File: SplunkRumWorkspace/SplunkRum/SplunkRum/RetryExporter.swift [hunk 2]
```diff
@@ -29,45 +29,41 @@ class RetryExporter: SpanExporter {
...
-        pending.append(contentsOf: spans)
-        if pending.count > MAX_PENDING_SPANS {
-            pending.removeFirst(pending.count - MAX_PENDING_SPANS)
-        }
-    }
-
-    func export(spans: [SpanData]) -> SpanExporterResultCode {
-        var result = attemptPendingExport()
+  func export(spans: [SpanData]) -> SpanExporterResultCode {
+        var result = attemptDBExport()
```
[the remark this answers is not in the document]
Comment (shattimare) [reply SplunkRumWorkspace/SplunkRum/SplunkRum/RetryExporter.swift:33]: When network is not available , we store all the spans generate during that time into DB. All this stored span will export to rum portal once network is available and any action or new span will be generated.

attemptDBExport function fetch span information from db and create new span using that information and export that spans. This function return success. In next line `result = proxy.export(spans: spans)` will send current span on playground.

[the remark this answers is not in the document]
Comment (seemk) [reply SplunkRumWorkspace/SplunkRum/SplunkRum/RetryExporter.swift:33]: I might be missing something, but I can't see where the spans loaded from disk are exported. It seems `attemptDBExport` fetches the spans, but does not export them anywhere?

[the remark this answers is not in the document]
Comment (shattimare) [reply SplunkRumWorkspace/SplunkRum/SplunkRum/RetryExporter.swift:33]: All db related stuffs are available in `CoreDataManager` class.In it there is a `fetchSpanValues` function, it  fetch span information from db and generate new span using  `createNewSpan` function, In `createNewSpan` function we use span builder to create new span using information from db and at the end we use `span.end(time: spanInfo.value(forKey: "duration") as! Date)` to end the span , as we end the span , span is added into the exporter queue(this logic is already developed in SDK) and network is available than it send to RUM playground using (Exporter’s chain) `ZipkinTraceExporter’s` export method.  If the span send successfully then delete that spans from db else not.

File: SplunkRumWorkspace/SplunkRum/SplunkRumTests/UtilsTests.swift [hunk 3]
```diff
@@ -196,12 +196,12 @@ class UtilsTests: XCTestCase {
...
         _ = re.export(spans: fiftySpans)
         _ = re.export(spans: fiftySpans)
-        XCTAssertEqual(0, localSpans.count)
+        XCTAssertEqual(150, localSpans.count)
 
         // now reconnected; next export of 50 should send total of 150 (last 100 retried)
         te.exportSucceeds = true
         _ = re.export(spans: fiftySpans)
-        XCTAssertEqual(150, localSpans.count)
+        XCTAssertEqual(250, localSpans.count)
```
Comment (seemk) [inline SplunkRumWorkspace/SplunkRum/SplunkRumTests/UtilsTests.swift:204]: Perhaps we need additional tests besides this?

File: SplunkRumWorkspace/SplunkRum/SplunkRum/DiskCache/CoreDataManager.swift [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+                span.setAttribute(key: dict.key, value: value as! Int)
+            } else if dict.value is Double {
+                span.setAttribute(key: dict.key, value: value as! Double)
+            } else if dict.value is Bool {
+                span.setAttribute(key: dict.key, value: value as! Bool)
+            } else if dict.value is [String: Any] {
+                span.setAttribute(key: dict.key, value: value as? AttributeValue)
+            }
+        }
+        span.end(time: spanInfo.value(forKey: "duration") as! Date)
```
Comment (seemk) [inline SplunkRumWorkspace/SplunkRum/SplunkRum/DiskCache/CoreDataManager.swift:130]: Is the `SpanContext` lost here?

Comment (shattimare) [reply SplunkRumWorkspace/SplunkRum/SplunkRum/DiskCache/CoreDataManager.swift:130]: @seemk, Here we are creating custom span using information stored in database, when we create custom span I think we are not using SpanContext.

File: SplunkRumWorkspace/SplunkRum/SplunkRumTests/UtilsTests.swift (hunk 3 above)
Comment (shattimare) [reply SplunkRumWorkspace/SplunkRum/SplunkRumTests/UtilsTests.swift:204]: @seemk added additional tests.

Comment (seemk) [reply SplunkRumWorkspace/SplunkRum/SplunkRumTests/UtilsTests.swift:204]: Can we verify that serialized attributes are all loaded correctly? I.e. when constructing spans from disk

Comment (shattimare) [reply SplunkRumWorkspace/SplunkRum/SplunkRumTests/UtilsTests.swift:204]: Add test case to check that attributes are loaded properly.

File: SplunkRumWorkspace/SplunkRum/SplunkRum/DiskCache/CoreDataManager.swift (hunk 4 above)
Comment (shattimare) [reply SplunkRumWorkspace/SplunkRum/SplunkRum/DiskCache/CoreDataManager.swift:130]: @seemk We can create SpanContext using traceId, spanID,t raceFlags, traceState. As mention in my very first  comment traceId, spanId, traceFlags are private property, So we can’t able to access it and create span context using that information stored on disk.

**Review by seemk — CHANGES_REQUESTED** [review]
There are a few problems with the current approach:
1. Spans are loaded from disk when a span is attempted to be exported. It would be better to do it continuously in the background.
2. There seems to be no limit to the amount of spans fetched - the spans loaded from disk will go through the same exporting loop, meaning the throttle will drop spans loaded from disk.

**Comment by shattimare** [comment]
> There are a few problems with the current approach:
> 
> 1. Spans are loaded from disk when a span is attempted to be exported. It would be better to do it continuously in the background.
> 2. There seems to be no limit to the amount of spans fetched - the spans loaded from disk will go through the same exporting loop, meaning the throttle will drop spans loaded from disk.

1. Implemented code for exporting cached span continuously when network is available. Did the code for it in `Networktype.swift` file.
2. Set limit of span fetching from disk, currently set it to 2 for testing purpose , we can change it later on.

**Comment by shattimare** [comment]
As per discussed with @seemk closing this PR.
he created separate branch and added this logic and merge to main branch.

**Pull request closed without merging**